node/impl/full/eth.go

```diff
@@ -754,6 +805,69 @@ func (a *EthModule) lookupEthAddress(ctx context.Context, addr address.Address)
 	return api.EthAddressFromFilecoinAddress(idAddr)
 }
 
+func (a *EthModule) newEthTxFromFilecoinMessage(ctx context.Context, smsg *types.SignedMessage) (api.EthTx, error) {
+	fromEthAddr, err := a.lookupEthAddress(ctx, smsg.Message.From)
+	if err != nil {
+		return api.EthTx{}, err
+	}
+
+	toEthAddr, err := a.lookupEthAddress(ctx, smsg.Message.To)
+	if err != nil {
+		return api.EthTx{}, err
+	}
+
+	toAddr := &toEthAddr
+	input := smsg.Message.Params
+	// Check to see if we need to decode as contract deployment.
+	// We don't need to resolve the to address, because there's only one form (an ID).
+	if smsg.Message.To == builtintypes.EthereumAddressManagerActorAddr {
+		switch smsg.Message.Method {
+		case builtintypes.MethodsEAM.Create:
+			toAddr = nil
+			var params eam.CreateParams
+			err = params.UnmarshalCBOR(bytes.NewReader(smsg.Message.Params))
+			input = params.Initcode
+		case builtintypes.MethodsEAM.Create2:
+			toAddr = nil
+			var params eam.Create2Params
+			err = params.UnmarshalCBOR(bytes.NewReader(smsg.Message.Params))
+			input = params.Initcode
+		}
+		if err != nil {
+			return api.EthTx{}, err
+		}
+	}
+	// Otherwise, try to decode as a cbor byte array.
+	// TODO: Actually check if this is an ethereum call. This code will work for demo purposes, but is not correct.
+	if toAddr != nil {
+		if decodedParams, err := cbg.ReadByteArray(bytes.NewReader(smsg.Message.Params), uint64(len(smsg.Message.Params))); err == nil {
+			input = decodedParams
+		}
+	}
+
+	r, s, v, err := api.RecoverSignature(smsg.Signature)
+	if err != nil {
+		return api.EthTx{}, err
+	}
```

Comment:
This will error out when dealing with non-Ethereum messages. Need to return 0 placeholders there.